### DISCUSSION TITLE
Resolve links/hardlinks for image scanning

### DISCUSSION THREAD

**Discussion by DmitriyLewen** [aquasecurity/trivy#5356] [Development]
We currently skip links when scanning images, because `archive/tar` package can't read value of link.
Typically hardlink and base file use same file name, and for us this is not problem because if we missed harklink - we will parse base file.
But 1 case was found:
`pnpm`can use hardlinks for `package.json` files from `node_modules/.pnpm/...` dir.
But base files doesn't use `package.json` filenames.
See this comment - https://github.com/aquasecurity/trivy/discussions/5332#discussioncomment-7238790.

To check links - we need to walk each tar archive (zipped layers) 2 times (to find links and read base files).
Trivy will take longer to work.

Perhaps we can add flag to enable link detection.

Welcome with out experiance, ideas, etc.

**Comment by simar7** [comment]
I think this is a wider issue that affect all kinds of scanning within Trivy. We also don't support the same in misconfiguration scanning. Please see this for context https://github.com/aquasecurity/trivy/issues/5002

cc @knqyf263

**Comment by DmitriyLewen** [comment]
`Conda` also uses hardlinks - #5360

**Reply by knqyf263** [reply]
@DmitriyLewen As I commented [here](https://github.com/aquasecurity/trivy/discussions/5360#discussioncomment-7311125), I thought hard links don't matter. They just look like normal files, don't they? It makes sense if you're talking about symlinks.

**Reply by DmitriyLewen** [reply]
I investigated this case in more detail.
Hard links make sense for `tar` package.
Looks like `tar` package uses same logic for hard and sym links.
tar.Reader is empty for hardlinks.

Some example:
```Dockerfile
FROM alpine

RUN mkdir /hard_test && echo "123" > /hard_test/foo && ln /hard_test/foo /hard_test/bar
```

I added the following logic [here](https://github.com/aquasecurity/trivy/blob/e2fb3dd58f2800ffbeb106ace8e30032c2600e1e/pkg/fanal/walker/tar.go#L55):
```go
if fileDir == "hard_test/" {
			fmt.Println(filePath)
			fmt.Println("filetype: " + string(hdr.Typeflag))
			b, _ := io.ReadAll(tr)
			fmt.Printf("lentght of reader: %d\n", len(b))
		}
```
result:
```
➜  ./trivy -d image 5356:test
...
hard_test/bar
filetype: 0 // TypeReg - https://cs.opensource.google/go/go/+/master:src/archive/tar/common.go;l=62
lentght of reader: 4
hard_test/foo
filetype: 1 // TypeLink - https://cs.opensource.google/go/go/+/master:src/archive/tar/common.go;l=68
lentght of reader: 0
...
```

**Reply by DmitriyLewen** [reply]
I thought this is related to Docker builds/layers/etc, but it appears to be feature of `tar` package:
1 more test:
```bash
➜ mkdir hard_test 
➜ echo "123" > hard_test/foo
➜ ln hard_test/foo hard_test/bar 
➜ tar -cf hard_test.tar hard_test 
```

Test project:
```go
package main

import (
	"archive/tar"
	"fmt"
	"io"
	"os"
)

func main() {
	f, err := os.Open("hard_test.tar")
	if err != nil {
		panic(err)
	}
	tr := tar.NewReader(f)
	for {
		hdr, err := tr.Next()
		if err == io.EOF {
			break
		} else if err != nil {
			panic(err)
		}
		fmt.Println(hdr.Name)
		fmt.Println("filetype: " + string(hdr.Typeflag))
		b, err := io.ReadAll(tr)
		if err != nil {
			panic(err)
		}
		fmt.Printf("lentght of reader: %d\n", len(b))
	}
}
```

Result:
```bash
➜  testproject go run main.go
hard_test/
filetype: 5
lentght of reader: 0
hard_test/foo
filetype: 0
lentght of reader: 4
hard_test/bar
filetype: 1
lentght of reader: 0

```

**Reply by knqyf263** [reply]
It is really interesting. Thanks for the update! The file type of symlink is also `TypeLink`? Don't we see any differences in symlink and hard link in tar?

**Reply by DmitriyLewen** [reply]
If i remember correctly - symlink uses [TypeSymlink](https://cs.opensource.google/go/go/+/master:src/archive/tar/common.go;bpv=1;bpt=1;l=69?gsn=TypeSymlink&gs=KYTHE%3A%2F%2Fkythe%3A%2F%2Fgo.googlesource.com%2Fgo%3Flang%3Dgo%3Fpath%3Darchive%2Ftar%23const%2520TypeSymlink)

**Reply by knqyf263** [reply]
I see. Can we resolve hard links? Does it have a file path to the original file?

**Reply by DmitriyLewen** [reply]
hardlink has filepath of original. Take a look screenshot in this comment - https://github.com/aquasecurity/trivy/discussions/5332#discussioncomment-7238790

**Reply by knqyf263** [reply]
Thanks. I was thinking about the implementation, but didn't develop a good idea. It looks stupid if we download the layer twice. Although we can download the layer to a temp file, the layer size can be a few GB. It consumes storage a lot.

**Reply by DmitriyLewen** [reply]
I was thinking about going through the layers twice:
1 - take all hardlinks (and symlinks if we need it) and save into map: main file - link
2 - It is our default run, but when we find file from map, we analyze this file and save it with several filepaths (i mean main filepath and filepaths for each links)

But it seems to take longer to scan.

**Comment by simar7** [comment]
Should we create an issue to track this @knqyf263?

**Reply by knqyf263** [reply]
@simar7 Please let me clarify the problem first. I'm still not understanding why hard links cause problems. If I understand correctly, the hard link doesn't have a base file. Hard links just refer to the same inode.

I think the problem with symbolic links is different from the one with hard links.

**Comment by manojkrishnanomula** [comment]
@knqyf263 did you conclude on how to make this approach because we're dependent on these hardlinks and symlinks as part of https://github.com/aquasecurity/trivy/discussions/5360

**Reply by knqyf263** [reply]
No, we haven't come up with a good idea yet. Please let us know if you have any ideas.
https://github.com/aquasecurity/trivy/discussions/5356#discussioncomment-7463144
